### PR TITLE
Release 0.20.1

### PR DESCRIPTION
**Pull request by Pat-Lafon**
This is an attempt to "kick-the-tires" on a new release based on https://github.com/orgs/lalrpop/discussions/823.

I did an initial draft of the release notes in `RELEASES.md` and ran `./version.sh` to update the version number in all the places.

### CODE REVIEW

File: RELEASES.md
Comment (dburgener) [inline RELEASES.md:6]: I might call this and the ascii space regex "bugfixes" rather than features.

File: RELEASES.md
Comment (dburgener) [inline RELEASES.md:15]: The phrase "in accordance to the MSRV policy" makes it sound like the increase is due to some policy rather than the is_terminal dependency requiring it.  Maybe just end at "MSRV increased to 1.70"?

**Review by dburgener — COMMENTED** [review]
Thanks for putting this together!  I think it looks mostly good from my perspective.  Just a couple of minor notes on the changelog.

**Comment by dburgener** [comment]
The notes look good, however I gave it one final test, and I'm seeing problems.  If I build using the commit before #814, everything works, but #814 without unicode support is translating my regexes containing `^` into an enumeration of all the possible unicode characters except excluded ones, and then panicing at run time upon encountering the unicode.  I'm looking into a root cause and fix, but I'm not sure if I'll be able to get to it tonight.

**Comment by dburgener** [comment]
I'm out of time for tonight unfortunately.  I'm getting weird failures both with and without unicode when trying to run tests in https://github.com/dburgener/cascade against lalrpop e2eb361cf507d5607f66c6b7ef4262d65770712d, but not against e2eb361cf507d5607f66c6b7ef4262d65770712d^.

When I run without unicode, I get the unexpected unicode on `^` patterns I mentioned above.   When I run with unicode, a few of my error tests for non-parsing input don't terminate in a reasonable amount of time.

I'm trying to get a simple reproducer that's more pure-lalrpop to narrow down what's wrong, but I haven't been successful at that yet.  It's not clear to me at this point why my code is failing but the lalrpop tests aren't.  Perhaps its user error on my end.  I'll take a look in the morning and follow up then.

File: doc/calculator/Cargo.toml
Comment (yannham) [inline doc/calculator/Cargo.toml:3]: Can't we use something like:

```
version.workspace = true
```

if this crate always follows the version of the main package of the workspace? This would save future version bumps.

File: doc/calculator/Cargo.toml
Comment (yannham) [inline doc/calculator/Cargo.toml:12]: Similarly, can we do something like:

```
lalrpop-util = { workspace = true, features = [ "lexer", "unicode" ] }
```

here, and put in the main `Cargo.toml` something like:

```
lalrpop-util = { version = "0.20.1", path = "./lalrpop-util"}
```

Sorry, it might be out of scope of this PR, but I've been doing that in other projects (define workspace-level dependencies for workspace members, and a workspace-level version, and use that everywhere else so that you only have to bump each member's, and only if it doesn't inherit the workspace's main version) and I wonder if it could make the workflow simpler for LALRPOP. Feel free to tell me that we can see that later :slightly_smiling_face:

File: version.sh
Comment (yannham) [inline version.sh:6]: Ah, I only see this script now. So maybe it's not really a hassle if it's automatized, although the scripts feel a bit ad-hoc.

**Review by yannham — COMMENTED** [review]
LGTM, beside side suggestions. I don't approve to wait for @dburgener to sort his tests out before potentially merging this.

File: doc/calculator/Cargo.toml
Comment (Pat-Lafon) [reply doc/calculator/Cargo.toml:3]: That's a fair point. It doesn't seem like any of these examples get rendered in the lalrpop book so it seems fine.

Actually on that point, maybe we should be pointing to these examples in the book instead of the tests when possible(referring to http://lalrpop.github.io/lalrpop/tutorial/index.html). Otherwise, I'm not sure how much visibility some of these get.

File: version.sh
Comment (Pat-Lafon) [reply version.sh:6]: It is somewhat automated, in a kinda hacky way. I think we could do a little bit better here in terms of setting up the release process steps. Not sure what big crates do in this regard. Maybe something that includes a mini-crater run or cargo-semvar-checks to help catch any last minute issues.

File: doc/calculator/Cargo.toml
Comment (Pat-Lafon) [reply doc/calculator/Cargo.toml:3]: I've spawned this off to it's own pr to be discussed/merged separately.

File: doc/calculator/Cargo.toml
Comment (Pat-Lafon) [reply doc/calculator/Cargo.toml:12]: See #843

**Comment by dburgener** [comment]
I *think* I have a solution for the issues I'm seeing.  I haven't tested it end to end yet, but limited testing I've done looks promising.

I believe the problem is that even before #814, when we use regex-syntax to parse our regexes, we didn't correctly disable unicode on non-unicode builds, which resulted in unicode regexes ending up in the generated parser.  Prior to #814 this seems to have not mattered, I *think* because internals of regex compile the unicode away anyways (based on [this](https://github.com/rust-lang/regex/issues/1088#issuecomment-1711727185) reply to a similar, but different issue.  The issue there was a bug in regex-syntax, but this is a bug in our code, that happens to hit a similar symptom).

Anyways, I'm going to tidy up my fix and do end to end testing.  If all goes well, I'll have a PR up in a bit.

**Comment by dburgener** [comment]
Oh, the other thing is that I'm still kind of mystified as to why our existing lalrpop tests didn't catch this at all.  I'm going to prioritize getting the fix in to unblock the release, but then I'll try to loop back and identify the testing gap.

**Comment by dburgener** [comment]
Okay, the second issue was user error in the sense of "why on earth would you do that, don't do that", and lalrpop error in the sense of "we should handle this bad input more sanely".

I had a skip pattern that could match zero length regexes.  #814 seems to have inadvertently dropped a check that returned "InvalidToken" if our longest match was a zero length skip pattern.

Note that the lalrpop book does recommend possibly zero length whitespace matches: https://lalrpop.github.io/lalrpop/lexer_tutorial/001_lexer_gen.html#customizing-skipping-between-tokens

I think the fix should be to restore the original check to return InvalidToken on a zero length skip match.  I'll submit a PR for that as well.

**Comment by Pat-Lafon** [comment]
All this sounds good to me! Thanks for the detailed look.

**Comment by dburgener** [comment]
https://github.com/lalrpop/lalrpop/pull/845 is up

Everything seems good as far as I can tell.  Once we get those merged, I'll retest.

**Comment by youknowone** [comment]
I'd like to add credits for new maintainers, who contributed a lot not only for this version but for longer time.

**Comment by youknowone** [comment]
is #844 the last piece of new release?

**Comment by dburgener** [comment]
> is https://github.com/lalrpop/lalrpop/pull/844 the last piece of new release?

That plus this PR of course is all I'm tracking on.  It looks like everyone is okay with #844, so I'll go ahead and merge that later today.

**Comment by dburgener** [comment]
#844 is merged.

I ran the same tests I did previous that uncovered the bugs against the latest master, and they passed this time.  This seems ready to go from a functional perspective in my opinion.

**Review by youknowone — APPROVED** [review]
Released 0.20.1 :+1:

**Comment by youknowone** [comment]
https://crates.io/crates/lalrpop/0.20.1

**Pull request merged**
